### ISSUE TITLE
Dropped A2DP Packets when sending stream to I2S interface for SPDIF (BT_APPL: Pkt dropped) (IDFGH-6738)

### ISSUE THREAD

**Issue by joba-1** [espressif/esp-idf#8368]
## Environment

- Development Kit:      [none] (ESP32 MH ET Live MiniKit)
- Module or chip used:  [ESP32-WROOM-32]
- IDF version (run ``git describe --tags`` to find it): v5.0-dev-1509-g5893797bf0
- Build System:         [idf.py]
- Compiler version (run ``xtensa-esp32-elf-gcc --version`` to find it): xtensa-esp32-elf-gcc (crosstool-NG esp-2021r2) 8.4.0
- Operating System:     [Linux]
- Using an IDE?: [No]
- Power Supply:         [USB]

## Problem Description

Scenario: using I2S to send a bluetooth A2DP stream (clean sine wave) as SPDIF to pin 27
Problem: The stream works for ~1,5 minutes, then starts to have a glitch every ~4-5s. 

The problem is not the added SPDIF encoding, the data coming from the esp callback is already wrong.

### Expected Behavior
The sine wave sent (or any other sound) gets transferred without continuity problems without time limit.

### Actual Behavior
The sine wave is perfect for ~1,5 minutes, then there are short intervals of ~30 samples not matching the sine wave.
After that the sine wave is good for ~4-5s, but not necessarily in phase to before the glitch.
Then the problem repeats until reset of the chip.

### Steps to reproduce

* Install latest ESP32 IDF (or any older - I also tried 4.4 and 3.5.5)
* get code with git clone git://github.com/amedes/esp_a2dp_sink_spdif.git
* run idf.py menuconfig to set output pin to 27 (optional if seeing the warning messages is enough)
* run idf.py flash monitor 
* connect phone via bluetooth to the ESP32 and send music (or better use a sine wave to see the glitch in the wave form)
* let it play for 1,5-2 min and start hearing soft clicks about every 4-5s
* watch the monitor to see "W (102105) BT_APPL: Pkt dropped" for each click.
* If you connect a red led to pin 27 and send the light into an optical SPDIF-In port of a soundcard you can record the wave directly with audacity.

// If possible, attach a picture of your setup/wiring here.

![image](https://user-images.githubusercontent.com/32450554/153273947-984a30ff-cc3a-454d-9dc1-986f59cf1211.png)

![image](https://user-images.githubusercontent.com/32450554/153273805-8fc8e66f-bcd4-4dfb-859e-161bd69925b5.png)

### Code to reproduce this issue
Full project (not mine): https://github.com/amedes/esp_a2dp_sink_spdif

The whole code is basically the esp_a2dp_sink example with added encoding of the spdif format and other I2S config for the increased bit frequency (to cover the SPDIF overhead). 

This is the function that initializes the I2S interface:

```cpp

// Some relevant defines for below function

#define I2S_NUM			(0)

#define I2S_BITS_PER_SAMPLE	(32)
#define I2S_CHANNELS		2
#define BMC_BITS_PER_SAMPLE	64
#define BMC_BITS_FACTOR		(BMC_BITS_PER_SAMPLE / I2S_BITS_PER_SAMPLE)
#define SPDIF_BLOCK_SAMPLES	192
#define SPDIF_BUF_DIV		2	// double buffering
#define DMA_BUF_COUNT		2
#define DMA_BUF_LEN		(SPDIF_BLOCK_SAMPLES * BMC_BITS_PER_SAMPLE / I2S_BITS_PER_SAMPLE / SPDIF_BUF_DIV)
#define I2S_BUG_MAGIC		(26 * 1000 * 1000)	// magic number for avoiding I2S bug
#define SPDIF_BLOCK_SIZE	(SPDIF_BLOCK_SAMPLES * (BMC_BITS_PER_SAMPLE/8) * I2S_CHANNELS)
#define SPDIF_BUF_SIZE		(SPDIF_BLOCK_SIZE / SPDIF_BUF_DIV)
#define SPDIF_BUF_ARRAY_SIZE	(SPDIF_BUF_SIZE / sizeof(uint32_t))

...

// initialize I2S for S/PDIF transmission
void spdif_init(int rate)
{
    int sample_rate = rate * BMC_BITS_FACTOR;
    int bclk = sample_rate * I2S_BITS_PER_SAMPLE * I2S_CHANNELS;
    int mclk = (I2S_BUG_MAGIC / bclk) * bclk; // use mclk for avoiding I2S bug
    i2s_config_t i2s_config = {
        .mode = I2S_MODE_MASTER | I2S_MODE_TX,
    	.sample_rate = sample_rate,
        .bits_per_sample = I2S_BITS_PER_SAMPLE,
        .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
        .communication_format = I2S_COMM_FORMAT_I2S,
        .intr_alloc_flags = 0,
        .dma_buf_count = DMA_BUF_COUNT,
        .dma_buf_len = DMA_BUF_LEN,
        .use_apll = true,
	.tx_desc_auto_clear = true,
    	.fixed_mclk = mclk,	// avoiding I2S bug
    };
    i2s_pin_config_t pin_config = {
        .bck_io_num = -1,
        .ws_io_num = -1,
        .data_out_num = SPDIF_DATA_PIN,
        .data_in_num = -1,
    };

    ESP_ERROR_CHECK(i2s_driver_install(I2S_NUM, &i2s_config, 0, NULL));
    ESP_ERROR_CHECK(i2s_set_pin(I2S_NUM, &pin_config));

    // initialize S/PDIF buffer
    spdif_buf_init();
    spdif_ptr = spdif_buf;
}
```

## Debug Logs

```
ets Jun  8 2016 00:22:57

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:6752
load:0x40078000,len:14796
load:0x40080400,len:3792
0x40080400: _init at ??:?

entry 0x40080694
I (27) boot: ESP-IDF v5.0-dev-1509-g5893797bf0 2nd stage bootloader
I (27) boot: compile time 19:11:56
I (27) boot: chip revision: 0
I (33) boot.esp32: SPI Speed      : 40MHz
I (36) boot.esp32: SPI Mode       : DIO
I (40) boot.esp32: SPI Flash Size : 2MB
I (45) boot: Enabling RNG early entropy source...
I (50) boot: Partition Table:
I (54) boot: ## Label            Usage          Type ST Offset   Length
I (61) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (69) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (76) boot:  2 factory          factory app      00 00 00010000 00100000
I (84) boot: End of partition table
I (88) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=1e388h (123784) map
I (141) esp_image: segment 1: paddr=0002e3b0 vaddr=3ffbdb60 size=01c68h (  7272) load
I (144) esp_image: segment 2: paddr=00030020 vaddr=400d0020 size=7c14ch (508236) map
I (331) esp_image: segment 3: paddr=000ac174 vaddr=3ffbf7c8 size=02bfch ( 11260) load
I (335) esp_image: segment 4: paddr=000aed78 vaddr=40080000 size=17dc0h ( 97728) load
I (377) esp_image: segment 5: paddr=000c6b40 vaddr=50000000 size=00010h (    16) load
I (389) boot: Loaded app from partition at offset 0x10000
I (389) boot: Disabling RNG early entropy source...
I (400) cpu_start: Pro cpu up.
I (401) cpu_start: Starting app cpu, entry point is 0x400812f8
0x400812f8: call_start_cpu1 at /home/joachim/esp/esp-idf/components/esp_system/port/cpu_start.c:152

I (0) cpu_start: App cpu up.
I (417) cpu_start: Pro cpu start user code
I (417) cpu_start: cpu freq: 160000000 Hz
I (417) cpu_start: Application information:
I (422) cpu_start: Project name:     esp_a2dp_sink_spdif
I (428) cpu_start: App version:      7fe5ed8-dirty
I (433) cpu_start: Compile time:     Feb  9 2022 19:26:44
I (439) cpu_start: ELF file SHA256:  dda7068dcc3408ea...
I (445) cpu_start: ESP-IDF:          v5.0-dev-1509-g5893797bf0
I (452) heap_init: Initializing. RAM available for dynamic allocation:
I (459) heap_init: At 3FFAFF10 len 000000F0 (0 KiB): DRAM
I (465) heap_init: At 3FFB6388 len 00001C78 (7 KiB): DRAM
I (471) heap_init: At 3FFB9A20 len 00004108 (16 KiB): DRAM
I (477) heap_init: At 3FFCA490 len 00015B70 (86 KiB): DRAM
I (483) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (490) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (496) heap_init: At 40097DC0 len 00008240 (32 KiB): IRAM
I (504) spi_flash: detected chip: gd
I (507) spi_flash: flash io: dio
W (511) spi_flash: Detected size(4096k) larger than the size in the binary image header(2048k). Using the size in the binary image header.
I (525) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.
I (535) gpio: GPIO[2]| InputEn: 0| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:0 
I (555) I2S: APLL expected frequency is 45158400 Hz, real frequency is 45158386 Hz
I (565) I2S: DMA Malloc info, datalen=blocksize=1536, dma_desc_num=2
I (565) I2S: I2S0, MCLK output by GPIO0
I (565) BTDM_INIT: BT controller compile version [33445ec]
I (575) system_api: Base MAC address is not set
I (575) system_api: read default base MAC address from EFUSE
I (585) phy_init: phy_version 4670,719f9f6,Feb 18 2021,17:07:07
W (1245) BT_BTC: A2DP Enable with AVRC
E (1255) BT_AV: Invalid A2DP event: 4
I (1265) BT_AV: event: 10
I (1265) BT_AV: event: 10
I (1265) BT_AV: event: 10
I (1265) BT_AV: event: 10
E (25345) BT_APPL: bta_av_rc_create ACP handle exist for shdl:0
I (25345) BT_AV: A2DP connection state: Connecting, [a0:cb:fd:33:11:19]
I (25375) BT_AV: A2DP audio stream configuration, codec type 0
I (25375) I2S: DMA queue destroyed
I (25385) I2S: APLL expected frequency is 45158400 Hz, real frequency is 45158386 Hz
I (25385) I2S: DMA Malloc info, datalen=blocksize=1536, dma_desc_num=2
I (25395) I2S: I2S0, MCLK output by GPIO0
I (25395) BT_AV: Configure audio player 21-15-2-35
I (25405) BT_AV: Audio player configured, sample rate=44100
W (25445) BT_APPL: new conn_srvc id:19, app_id:0
I (25445) BT_AV: A2DP connection state: Connected, [a0:cb:fd:33:11:19]
W (25455) BT_BTC: Function btc_avrc_ct_send_set_absolute_volume_cmd() called when RC is not connected
I (25525) RCCT: AVRC conn_state evt: state 1, [a0:cb:fd:33:11:19]
I (25535) RCTG: AVRC conn_state evt: state 1, [a0:cb:fd:33:11:19]
I (25535) RCCT: AVRC remote features 25b, TG features 51
I (25545) RCTG: AVRC remote features 25b, CT features 2
I (25595) RCCT: remote rn_cap: count 7, bitmask 0xf26
I (25605) RCTG: AVRC register event notification: 13, param: 0x0
I (25745) RCCT: AVRC metadata rsp: attribute id 0x1, sine-441-661
I (25745) RCCT: AVRC metadata rsp: attribute id 0x2, joba-1
I (25745) RCCT: AVRC metadata rsp: attribute id 0x4, sines
I (25745) RCCT: AVRC metadata rsp: attribute id 0x20, Test
I (28675) BT_AV: event: 13
I (31685) BT_AV: event: 13
I (31695) BT_LOG: bta_av_link_role_ok hndl:x41 role:1 conn_audio:x1 bits:1 features:x824b

W (31705) BT_APPL: new conn_srvc id:19, app_id:1
I (31705) RCCT: AVRC event notification: 1
I (31705) BT_AV: Playback status changed: 0x1
I (31715) BT_AV: A2DP audio state: Started
I (31715) RCCT: AVRC event notification: 5
I (31725) BT_AV: Play position changed: 53018-ms
I (33225) BT_AV: Audio packet count 100
I (34805) BT_AV: Audio packet count 200
I (36265) BT_AV: Audio packet count 300
I (37705) BT_AV: Audio packet count 400
I (37885) RCCT: AVRC event notification: 5
I (37885) BT_AV: Play position changed: 22-ms
I (39185) BT_AV: Audio packet count 500
I (40675) BT_AV: Audio packet count 600
I (42085) BT_AV: Audio packet count 700
I (43565) BT_AV: Audio packet count 800
I (44975) BT_AV: Audio packet count 900
I (46445) BT_AV: Audio packet count 1000
I (47935) BT_AV: Audio packet count 1100
I (47955) RCCT: AVRC event notification: 5
I (47955) BT_AV: Play position changed: 10091-ms
I (49265) RCTG: AVRC set absolute volume: 7%
I (49275) RCTG: Volume is set by remote controller 7%

I (49355) BT_AV: Audio packet count 1200
I (49735) RCTG: AVRC set absolute volume: 13%
I (49735) RCTG: Volume is set by remote controller 13%

I (49925) RCTG: AVRC set absolute volume: 26%
I (49925) RCTG: Volume is set by remote controller 26%

I (50055) RCTG: AVRC set absolute volume: 53%
I (50055) RCTG: Volume is set by remote controller 53%

I (50185) RCTG: AVRC set absolute volume: 74%
I (50185) RCTG: Volume is set by remote controller 74%

I (50375) RCTG: AVRC set absolute volume: 93%
I (50375) RCTG: Volume is set by remote controller 93%

I (50475) RCTG: AVRC set absolute volume: 100%
I (50475) RCTG: Volume is set by remote controller 100%

I (50845) BT_AV: Audio packet count 1300
I (52305) BT_AV: Audio packet count 1400
I (53735) BT_AV: Audio packet count 1500
I (55195) BT_AV: Audio packet count 1600
I (56695) BT_AV: Audio packet count 1700
I (58115) BT_AV: Audio packet count 1800
I (58125) RCCT: AVRC event notification: 5
I (58125) BT_AV: Play position changed: 20194-ms
I (59595) BT_AV: Audio packet count 1900
I (61015) BT_AV: Audio packet count 2000
I (62485) BT_AV: Audio packet count 2100
I (63985) BT_AV: Audio packet count 2200
I (65395) BT_AV: Audio packet count 2300
I (66875) BT_AV: Audio packet count 2400
I (68185) RCCT: AVRC event notification: 5
I (68185) BT_AV: Play position changed: 30340-ms
I (68355) BT_AV: Audio packet count 2500
I (69765) BT_AV: Audio packet count 2600
I (71215) BT_AV: Audio packet count 2700
I (72675) BT_AV: Audio packet count 2800
I (74155) BT_AV: Audio packet count 2900
I (75595) BT_AV: Audio packet count 3000
I (77045) BT_AV: Audio packet count 3100
I (78345) RCCT: AVRC event notification: 5
I (78355) BT_AV: Play position changed: 40421-ms
I (78525) BT_AV: Audio packet count 3200
I (80015) BT_AV: Audio packet count 3300
I (81425) BT_AV: Audio packet count 3400
I (82875) BT_AV: Audio packet count 3500
I (84385) BT_AV: Audio packet count 3600
I (85795) BT_AV: Audio packet count 3700
I (87295) BT_AV: Audio packet count 3800
I (88685) RCCT: AVRC event notification: 5
I (88685) BT_AV: Play position changed: 50560-ms
I (88715) BT_AV: Audio packet count 3900
I (90175) BT_AV: Audio packet count 4000
I (91645) BT_AV: Audio packet count 4100
I (93065) BT_AV: Audio packet count 4200
I (94545) BT_AV: Audio packet count 4300
I (95995) BT_AV: Audio packet count 4400
I (97455) BT_AV: Audio packet count 4500
I (97635) RCCT: AVRC event notification: 5
I (97635) BT_AV: Play position changed: 29-ms
I (98905) BT_AV: Audio packet count 4600
I (100415) BT_AV: Audio packet count 4700
I (101835) BT_AV: Audio packet count 4800
W (102105) BT_APPL: Pkt dropped

I (103315) BT_AV: Audio packet count 4900
I (104725) BT_AV: Audio packet count 5000
I (106205) BT_AV: Audio packet count 5100
W (106765) BT_APPL: Pkt dropped

I (107685) BT_AV: Audio packet count 5200
I (107695) RCCT: AVRC event notification: 5
I (107705) BT_AV: Play position changed: 10103-ms
I (109095) BT_AV: Audio packet count 5300
W (110115) BT_APPL: Pkt dropped

I (110595) BT_AV: Audio packet count 5400
I (112075) BT_AV: Audio packet count 5500
W (112995) BT_APPL: Pkt dropped

I (113485) BT_AV: Audio packet count 5600
I (114965) BT_AV: Audio packet count 5700
I (116445) BT_AV: Audio packet count 5800
I (117855) BT_AV: Audio packet count 5900
I (117855) RCCT: AVRC event notification: 5
I (117855) BT_AV: Play position changed: 20203-ms
W (118135) BT_APPL: Pkt dropped

I (119355) BT_AV: Audio packet count 6000
W (120735) BT_APPL: Pkt dropped

I (120755) BT_AV: Audio packet count 6100
...
```

## Other items if possible

- [ ] sdkconfig file (attach the sdkconfig file from your project folder)
[sdkconfig.txt](https://github.com/espressif/esp-idf/files/8035644/sdkconfig.txt)

- [ ] elf file in the ``build`` folder (**note this may contain all the code details and symbols of your project.**)
[esp_a2dp_sink_spdif.zip](https://github.com/espressif/esp-idf/files/8035652/esp_a2dp_sink_spdif.zip)

- [ ] coredump (This provides stacks of tasks.)
none

**Comment by joba-1** [comment]
If I send slightly less data than the callback delivers, the drops are gone.

**Comment by BetterJincheng** [comment]
@joba-1 I played music without any pkt dropping with your code and config. Any suggestions? Thanks!

**Comment by BetterJincheng** [comment]
@joba-1 And I did not understand how your 'rate control' works...

**Comment by BetterJincheng** [comment]
@joba-1 
```
size_t write_ringbuf(const uint8_t *data, size_t size)
{
    // rate control
    UBaseType_t items;
    vRingbufferGetInfo(s_ringbuf_i2s, NULL, NULL, NULL, NULL, &items);
    if (items < RINGBUF_SIZE * 3 / 8) {
        xRingbufferSend(s_ringbuf_i2s, (void *)data, AUDIO_SAMPLE_SIZE, (portTickType)portMAX_DELAY);
    } else if (items > RINGBUF_SIZE * 5 / 8) {
        size -= AUDIO_SAMPLE_SIZE;
    }

    BaseType_t done = xRingbufferSend(s_ringbuf_i2s, (void *)data, size, (portTickType)portMAX_DELAY);
    if(done){
        return size;
    } else {
        return 0;
    }
}
```

**Comment by joba-1** [comment]
Ah, I see now what you mean. I wasn't aware of this code.

This rate limit code does the following:

* If the ring buffer is rather empty, then send the first audio sample of the callback twice
* If the ring buffer is rather full, then discard the last audio sample of the callback

It assumes, the rate bt sends data and i2s consumes data is out of sync for not more than one sample per callback.
I'll add some logs and try to drop/add more samples if the buffer is even closer to full/empty and hope this is still not audible.

Looks promising, thanks for the hint.

**Comment by joba-1** [comment]
> @joba-1 I played music without any pkt dropping with your code and config. Any suggestions? Thanks!

I can only assume that your bt sender is a tiny bit slower than mine or my esp is running a bit slower than yours

**Comment by joba-1** [comment]
dropped even more frames to a maximum of 100 out of 2560 bytes, then drops are gone.

**Comment by BetterJincheng** [comment]
@joba-1 "Pkt dropped" happens when sink_rx_queue is about full. The root cause is that the Bluetooth Stack is busy to handle 
the incoming audio data. 

Now some information for your reference.

The task priority of `bt_app_a2d_data_cb` is 19, the priority of task handling sink_rx_queue is 20, and, your  I2S task priority is 22 .

A possible case is that, your I2S is always busy, `bt_app_a2d_data_cb` has fewer chance to occupy the CPU. So packet dropped can happened.

Lower down the priority of I2S task to 18, and try again.

Turn to us anytime.
Thanks

**Comment by joba-1** [comment]
thanks for the update. Will try the prio change.

**Comment by AxelLin** [comment]
> Lower down the priority of I2S task to 18, and try again.
> 
> Turn to us anytime. Thanks

@BetterJincheng 
FYI, there is a bug regarding task priority if using esp32 dual-core.
https://github.com/espressif/esp-idf/issues/7580

**Comment by BetterJincheng** [comment]
@joba-1 Sometimes, there are misunderstandings about the use of `data_cb`. The really reasonable thing to do is， copy the data and let the `data_cb` returns, handle the data in another task.

**Comment by joba-1** [comment]
I can use any guidance, thanks for that.

Currently I (or better: amedes) do it like in the original espressif a2dp to i2s example: the cb puts data into a circular buffer and another task fetches it from there.
This other task converts the samples to spdif format (which doubles the sample size) and writes that out with double speed i2s to a data pin (that drives the toslink led).

My main problem seems to be not priority, but that a2dp bit rate (which is controlled solely by the external bt device) and the spdif i2s bit rate need to be precisely synchronous but are not. 
The current code adds or drops at most 1 sample per cb (~1/1000th), but I require to drop much more (>10 samples per cb). That way the circular buffer fills up and then starts blocking the cb, no matter what prio it has.
Of course I cannot statically drop e.g. 20 frames each cb. If the number is too high I would drain the buffer sooner or later and get ugly sound artifacts as well.

Is it possible to reduce the packet size I get from the bt cb? Else I try to split it up in an intermediate level.

**Comment by joba-1** [comment]
I tried increasing (and decreasing) the task prios. Did not help.
Then I rewrote the [rate limiting code](https://github.com/joba-1/esp_a2dp_sink_spdif/blob/c7d33f29a52e7da59f4baee206bcc0fad7681f7e/main/bt_app_core.c#L192) to use a floating average buffer usage instead of the current usage (fluctuated too much) and allowed dropping up to 3 samples. That did it. 

Thanks for your support, your hint on the rate limiting code did it!
